addons/website_forum/populate/mail_message.py

```diff
@@ -0,0 +1,40 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import models
+from odoo.tools import populate
+
+CP_WEIGHTS = { 1: 35, 2: 30, 3: 25, 4: 10 }
+
+class Message(models.Model):
+    _inherit = "mail.message"
+    _populate_dependencies = ["forum.post"]
+    _populate_sizes = {'small': 100, 'medium': 1000, 'large': 90000 }
+
+    def _populate_factories(self):
+        question_ids = self.env.registry.populated_models['forum.post']
+        answers = self.env['forum.post'].search_read([('parent_id', 'in', question_ids)], ['id'])
+        answer_ids = [id for answer in answers for id in answer.values()]
+        nb_comments = 0
+        is_question = True
+        post_to_comment_id = None
+
+        def get_res_id(random=None, **kwargs):
+            nonlocal nb_comments, is_question, post_to_comment_id
+            if nb_comments == 0:
+                nb_comments = random.choices(*zip(*CP_WEIGHTS.items()))[0]
+                if is_question and question_ids:
+                    post_to_comment_id = question_ids.pop()
+                    is_question = False
+                elif answer_ids:
+                    post_to_comment_id = answer_ids.pop()
+                    is_question = True
+            nb_comments -= 1
+            return post_to_comment_id
```

Comment:
How about making this a generator, like above ? That way, you can keep the variables in the scope of the function.
(Other comments not included here)

```suggestion

        def compute_res_id(iterator, **kwargs):
            nb_comments = 0
            is_question = True
            post_to_comment_id = None

            while True:
                if nb_comments == 0:
                    nb_comments = random.choices(*zip(*CP_WEIGHTS.items()))[0]
                    if is_question and question_ids:
                        post_to_comment_id = question_ids.pop()
                        is_question = False
                    elif answer_ids:
                        post_to_comment_id = answer_ids.pop()
                        is_question = True
                nb_comments -= 1
                yield post_to_comment_id
```